dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/relation/RelationModelManager.java

```diff
@@ -0,0 +1,577 @@
+        }
+        return result;
+    }
+
+    private RelationModel computeUnion(Relation r, List<RelationModel> relations) {
+        RelationModel result = newModel(r);
+        for (RelationModel rm : relations) {
+            result.addEdges(rm.getEdges());
+        }
+        return result;
+    }
+
+    private RelationModel computeIntersection(Relation r, List<RelationModel> relations) {
+        RelationModel result = newModel(r);
+        if (relations.size() == 1) {
+            result.addEdges(relations.get(0).getEdges());
+        } else {
+            if (relations.size() == 2) {
+                Set<EdgeModel> edges = Sets.intersection(relations.get(0).getEdges(), relations.get(1).getEdges());
+                if (relations.size() > 2) {
+                    for (int i = 2; i < relations.size(); i++) {
+                        edges = Sets.intersection(edges, relations.get(i).getEdges());
+                    }
+                }
+                result.addEdges(edges);
+            }
+        }
+        return result;
+    }
+
+    private RelationModel computeTransitiveClosure(Relation r, RelationModel rm) {
+        RelationModel result = newModel(r);
+        final Map<EventData, Set<EventData>> reachMap = new HashMap<>();
+        for (EdgeModel edge : rm.getEdges()) {
+            reachMap.computeIfAbsent(edge.getPredecessor(), k -> new HashSet<>()).add(edge.getSuccessor());
+        }
+        for (EventData p : reachMap.keySet()) {
+            final Set<EventData> reachables = new HashSet<>();
+            depthFirstSearch(p, reachMap, reachables);
+            for (EventData s : reachables) {
+                if (p != s) { addEdgeToRelation(result, p, s); }
+            }
+        }
+        return result;
+    }
+
+    private void depthFirstSearch(EventData p, Map<EventData, Set<EventData>> reachMap, Set<EventData> visited) {
+        if (visited.contains(p)) { return; }
+        visited.add(p);
+        if (reachMap.containsKey(p)) {
+            for (EventData s : reachMap.get(p)) {
+                depthFirstSearch(s, reachMap, visited);
+            }
+        }
+    }
+
+    private RelationModel removeSelfLoopToShow(RelationModel rm) {
+        for (EdgeModel e : rm.getEdges()) {
+            if (e.getPredecessor() == e.getSuccessor()) {
+                rm.removeEdgeToShow(e);
+            }
+        }
+        return rm;
+    }
+
+    // We do NOT show transitive edges in base relations.
+    private RelationModel removeTransitiveEdgesToShow(RelationModel rm) {
+        rm = removeSelfLoopToShow(rm);
+        for (EdgeModel e1 : rm.getEdges()) {
+            for (EdgeModel e2 : rm.getEdges()) {
+                if (e1.getSuccessor() == e2.getPredecessor()) {
+                    rm.removeEdgeToShow(new EdgeModel(e1.getPredecessor(), e2.getSuccessor()));
+                }
+            }
+        }
+        return rm;
+    }
+
+
+    private final class RelationModelBuilder implements Visitor<RelationModel> {
```

Comment:
I would advice against reimplementing the computation of derived relations: this is very tricky.
Indeed, your code fails on recursive memory models and fixing this is non-trivial.

You should take a look into the package of `CAATSolver`, especially into the classes `PredicateHierarchy` and `SimpleGraph`.
A `PredicateHierarchy` has a concrete set of base predicates (~ base relations) and a set of derived predicates (~ derived relations), similar to `Wmm`. What you can then do is populate the base graphs with concrete edges (your `RelationModels`) and then let the `PredicateHierarchy` correctly populate all derived graphs.
Constructing a `PredicateHierarchy` from a `Wmm` has some minor edge cases, but you can check out `ExecutionGraph` which does precisely this.
However, that class uses custom graph implementations for the base predicates that can populate themselves from an SMT model.
In your case, I would just copy that code but replace all these special graphs for base relations by `SimpleGraph`, which you can populate however you want (e.g., with edges from your `RelationModel` of the base relations).

Reply:
Thanks for the comment.

>fails on recursive memory models

Could you please give me an example of this?

Reply:
Yes, this does not work ... there seems to be infinite recursive calls to 
```
com.dat3m.dartagnan.verification.model.relation.RelationModelManager.extractRelation(RelationModelManager.java:68)
```

Reply:
>In your case, I would just copy that code but replace all these special graphs for base relations by SimpleGraph, which you can populate however you want (e.g., with edges from your RelationModel of the base relations).

Hey @ThomasHaas , thanks for the advice. For computation of derived relations, what about using the specific graphs also for each base relation and let them populate themselves? So we will have a consistent way on the relation computation for solving and witness generation, the only difference is that for latter we do not construct all relations but perform the computation on demand. 

And for special requirements for witness, corresponding methods can be implemented to populate the `RelationModels`. E.g. in the graph we want `PO` edges to connect also `Locals` and `Asserts`.

Reply:
That won't work out-of-the-box, because the specific graphs populate themselves based on the `ExecutionModel` or the underlying `(SMT)Model`. In this case, the `ExecutionModel` already needs to contain enough information for all base relations, i.e., it already needs `rf`, `co`, etc.
In other words, they populate themselves based on "implicit" `RelationModels` so you cannot generate the `RelationModels` from the special graphs without running into cycles :). Btw. the special graphs not only populate themselves but also rely on implicit representation, e.g., the `ProgramOrderGraph` uses only `O(1)` space to describe a relation of size `O(n^2)`.
Now, it is true that e.g. `ProgramOrderGraph` can populate itself based solely on `EventData` and that could be reused, but at least for `rf` and `co` you need to do it differently.

Let me summarize what I think you should do:
(1) Have explicit(*) graphs for the base relations, i.e., `SimpleGraph`. This is as general as it gets because those explicit graphs can be generated from the SMT model, from parsing files, or even by hand.
(2) Reuse `CAAT`-functionality to compute all derived relations on demand.
(3) Do NOT add `Locals` and `Asserts` into `po`. This is incorrect and will cause all derived relations to be messed up. Instead, you should either have a special `localPo` relation model that you maintain but that is never used in the derived relations (well, you could compute `po` via `let po = localPo & _*_` if you wanted), or you simply change the printing of witnesses to include the local events without relying on `po` to do so.

To elaborate more on point (3) above. I think your `ExecutionModel` should have `EventData/Models` for every piece of executed code (maybe modulo minor exceptions). But no matter what events it contains, the base relations used in the memory model should never contain any of the extra events (except maybe special internal relations like `idd`, if you even want to track them at all).
Also, just because the `ExecutionModel` contains information about events/relations, the printer does not necessarily have to print them! In particular, you should not force events into relations or omit events from the model just because you do/don't want to print them in the end. I can easily imagine that the user could specify levels-of-detail when printing the model.

(*) Well, for relations like `po`, `int`, `ext` and `loc` you could also have implicit representations based on the `EventData/Model` to save space (this is what CAAT does). But for now, it is easiest to just make all of them explicit.

Reply:
OK, understood. 

Eventually, the `ExecutionModel` should be:

(1) in the initialization, `EventModels` for all `Events` and `RelationModels` for at least `po`, `rf`, `co` should be constructed, which provides the necessary information for the solving purpose of `CAAT`-part, i.e. the special graphs can populate themselves and the `PredicateHierarchy` can perform the computation.
(2) for witness generation, more constructions of `RelationModels` and computations of derived ones happen on demand.

Am I right about this part?

BTW, my plan is to implement the `RelationModel` first and let this model interact perfectly with other parts of what we have now, which is also the goal of this PR. And the next step is to implement the `EventModel` and let it "replace" the current `EventData` and also adjust the `ExecutionModel` as what we want it to be (like discussed in #732 )

Reply:
There is a question about efficiency here. If you always generate explicit models for `po`, `rf`, `co`, etc., you can also replace CAAT's special graphs and let it directly work over your explicit representation. The disadvantage, of course, is that this is theoretically more space/time-consuming. I can imagine that it does not matter much though.

But maybe, you should tackle this problem a little bit differently to avoid interacting with the CAAT-based solving at all for now (this makes things just more difficult).
You could keep the existing `ExecutionModel` class for use by CAAT (it was originally written with CAAT in mind), but introduce a new one for all other usages (mainly printing right now). There you can write all the code from scratch without worrying that it needs to work with existing infrastrucutre. Once you have a clean implementation of `(New)ExecutionModel`, `EventModel` and `RelationModel`, you can start replacing existing usages of the old `ExecutionModel` by the new one step-by-step, starting from printing-related code to eventually CAAT.
The point is that we first need a way to represent all the information of an execution we could possibly care about, and then we need to see how we can make use of it.